pkg/kubelet/metrics/metrics.go

```diff
@@ -106,20 +115,33 @@ var (
 			Subsystem:      KubeletSubsystem,
 			Name:           PodWorkerDurationKey,
 			Help:           "Duration in seconds to sync a single pod. Broken down by operation type: create, update, or sync",
-			Buckets:        metrics.DefBuckets,
+			Buckets:        podLatencyBuckets,
 			StabilityLevel: metrics.ALPHA,
 		},
 		[]string{"operation_type"},
 	)
 	// PodStartDuration is a Histogram that tracks the duration (in seconds) it takes for a single pod to go from pending to running.
-	PodStartDuration = metrics.NewHistogram(
+	PodStartDuration = metrics.NewHistogramVec(
 		&metrics.HistogramOpts{
 			Subsystem:      KubeletSubsystem,
 			Name:           PodStartDurationKey,
 			Help:           "Duration in seconds for a single pod to go from pending to running.",
-			Buckets:        metrics.DefBuckets,
+			Buckets:        podLatencyBuckets,
 			StabilityLevel: metrics.ALPHA,
 		},
+		[]string{"namespace"},
+	)
+
+	// PodStartE2EDuration is a histogram that tracks duration in seconds for a single pod to go from created to running
```

Comment:
Just comparing the descriptions of PodStartDuration and PodStartE2EDuration, it is hard to tell what the difference is, since pods are pending when they are first created.  It may be useful to update the text for PodStartDuration to indicate that it is the time since the kubelet received the pod to distinguish the metrics.

Reply:
is changing metrics name acceptable?

or I can rename the new one as "pod_create_to_start_duration".

Reply:
changing metric name is not acceptable.  It is the equivalent of removing the metric and adding a new one.  If we think it is neccessary, we can migrate from old to new by having both for a few releases, but that is a last-resort change.

I was suggesting just updating the help text.  I would probably give the new one a different name.  `pod_creation_duration_seconds` seems reasonable, but feel free to choose something else.  As long as the help text is explicit about what start and end times are (e.g. pod object creation to status update to running vs. kubelet begins the process of creating a pod vs ends the process of creating a pod).

I can also see a case for this metric being measured from outside the kubelet (e.g. kube-state-metrics), rather than inside.  It is possible to do so, since both the creation time, and "Ready" condition's last transition time can be observed outside of the kubelet.  I'm also curious what kubelet problems would be caught by this metric that wouldn't be caught by the `pod_start_duration_seconds` metric

Reply:
WDYT about the last suggestion?  I think it is strange to measure the latency from api object creation to when the kubelet observes from the container runtime that the pod is ready.  I would expect either an API-to-API e2e latency (probably measured from outside the kubelet), or a kubelet-action-to-kubelet-observation latency.

Reply:
@dashpole hi David, sry to have this slip away. 

In practice, especially with cluster autoscaling, this provides application developers very useful insights about infra level SLA, I.e, after I submit my workload, this is the heuristics about how soon my code starts to run, given all factors in place, scheduling (mostly queuing) delay, or wait for new compute resources to spin up, or to pull image.

We tried to record it in other components but finally realized that kubelet is the only place to make things accurate as it knows exactly the first time pod starts to run. Also pod status update can have significant enough delays.

Reply:
Is the point at which the pod starts to run really the point we care about?  There are other "infra" pieces outside of the kubelet that are part of making sure pods are completely set-up.  See https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/0007-pod-ready%2B%2B.md, which is about incorporating service readiness into the pod status.  From controller's (e.g. deployment controller) perspective, the pod is only really running once the status has been updated to Ready.

I agree that status updates can be slow, but I don't think pod creation to containers running is the metric we should use.  IMO we should measure a few different things:

1. scheduling latency (scheduler receives pod to pod assigned to node)
2. Kubelet start pod latency (kubelet receives pod to containers running)
3. E2e latency (pod object created to pod object running)

If you just want to measure component health (e.g. make a kubelet/scheduler SLA), you should use 1, and 2.  If you want to measure when the pod is ready, you should use 3 even though it includes status manager updates.

Reply:
gotcha, I think your summary makes sense to me. I will proceed to remove this metric. Anything else you'd like me to make change? @dashpole

Reply:
Left one other comment below.